frostsnap_core/src/coordinator.rs

```diff
@@ -1174,93 +1182,192 @@ impl FrostCoordinator {
+                if let Some(access_structure_ref) = restoration.access_structure_ref {
+                    if access_structure_ref != recover_share.held_share.access_structure_ref {
+                        return Err(RestoreRecoverShareError::AcccessStructureMismatch);
+                    }
+                }
+
+                if restoration.key_name != recover_share.held_share.key_name {
+                    return Err(RestoreRecoverShareError::NameMismatch);
+                }
+            }
+            None => return Err(RestoreRecoverShareError::UnknownRestorationId),
         }
 
         Ok(())
     }
 
-    pub fn recover_access_structure(
+    pub fn finish_restoring(
         &mut self,
-        access_structure_ref: AccessStructureRef,
+        restoration_id: RestorationId,
         encryption_key: SymmetricKey,
         rng: &mut impl rand_core::RngCore,
-    ) -> Result<(), ActionError> {
-        let key = self
-            .keys
-            .get(&access_structure_ref.key_id)
-            .ok_or(ActionError::StateInconsistent("key doesn't exist".into()))?;
-        let pending_as = key
-            .recovering_access_structures
-            .get(&access_structure_ref.access_structure_id)
-            .ok_or(ActionError::StateInconsistent(format!(
-                "access structure for recovery not found {:?}",
-                access_structure_ref
-            )))?;
-        if pending_as.share_images.len() >= pending_as.threshold as usize {
-            let share_images = pending_as
-                .share_images
-                .clone()
-                .into_iter()
-                .map(|(share_index, (_, share_image))| (share_index, share_image))
-                .collect::<Vec<_>>();
-            let implied_root_poly = poly::point::interpolate(&share_images);
-            let implied_root_poly = poly::point::normalize(implied_root_poly).collect::<Vec<_>>();
-            let root_shared_key = SharedKey::from_poly(implied_root_poly).non_zero().ok_or(
-                ActionError::StateInconsistent("can't recover a zero key".into()),
-            )?;
-            let device_to_share_index = pending_as
-                .clone()
-                .share_images
-                .into_iter()
-                .map(|(party_index, (device_id, _))| (device_id, party_index))
-                .collect();
-            self.mutate_new_key(
-                key.key_name.clone(),
-                root_shared_key,
-                device_to_share_index,
-                encryption_key,
-                key.purpose,
-                rng,
-            );
+    ) -> Result<AccessStructureRef, RestorationError> {
+        let state = self
+            .restorations
+            .get(&restoration_id)
+            .ok_or(RestorationError::UnknownRestorationId)?;
+
+        let root_shared_key = state
+            .clone()
+            .access_structure
+            .interpolate()
+            .ok_or(RestorationError::NotEnoughShares)?;
```

Comment:
In this design it looks like we can recover with >T shares, which in this case, this interpolate and underlying `from_share_images` doesn't detect/handle the case that one of them is incompatible. Ties into the wider design question of how we approach incompatible backups.

Reply:
It does handle it. If one is incompatible the threshold won't be what was expected.